### DISCUSSION TITLE
Polyfill PHP80 causing classname collision

### DISCUSSION THREAD

**Discussion by slikk66** [symfony/symfony#45561] [Q&A] [answered]
Hi, we have a class in our system named Attribute.  We pulled in the googleads package with composer, and it pulls in polyfill-php80 which also has an Attribute class       **_with no namespace attached_**.

This seems to be counter to the point of composer and packages, where the packages included should be namespaced properly to avoid causing issues inside the app they're brought into.

From what I can tell, this polyfill is some type of "hack" basically to allow programs to use php 8 syntax in versions below 8.  Is that correct?

We ended up just renaming our class and all the calls to it inside of our application, which just feels gross.  But wasn't sure what else to do.

Any thoughts/help?

Thank you

**Comment by ghost** [comment]
I would say you should not have no namespaced class called `Attribute` when trying to polyfill PHP 8.0 features.

P.S. I wonder why this class is not mentioned in the README file and why not inside `if (\PHP_VERSION_ID < 80000)` statement as the rest of the classes.

**Comment by slikk66** [comment]
We are using php 7.4 currently.  We did not include polyfill-php80 on our own accord,  Google Ads package did. It broke our app, so it probably should at minimum have a list of reserved class names.

And I suppose packages that use it should mention it specifically as a possible issue (harder to do that, I'm sure).

Just wanted to bring it up as it caused an issue in production we didn't catch during testing as only happened on a deep link page at runtime.

**Comment by craigh** [comment]
Google Ads (https://packagist.org/packages/googleads/googleads-php-lib) does not appear to require https://github.com/symfony/polyfill-php80

**Comment by slikk66** [comment]
This is the one we added https://packagist.org/packages/googleads/google-ads-php

**Comment by mbabker** [comment, marked as answer]
> We ended up just renaming our class and all the calls to it inside of our application, which just feels gross. But wasn't sure what else to do.

This is the right thing to do.  PHP 8 adds an `Attribute` class in the root (global) namespace, so you would have needed to rename your class anyway.  The polyfill package being installed just forced you to do it sooner.

You can use the `composer why` command to determine what is requiring any package (i.e. `composer why symfony/polyfill-php80` to see what requires the PHP 8.0 polyfill).

Bug reports and/or pull requests (such as one adding the `Attribute` class to the PHP 8.0 package's README) for all polyfill packages can be logged to https://github.com/symfony/polyfill.

**Reply by slikk66** [reply]
Thanks Michael, selected as answer.
